File: Leetcode/200. Number of Islands/Solution.java
/* https://leetcode.com/problems/number-of-islands/description/

Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. 
An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. 
You may assume all four edges of the grid are all surrounded by water.

Example 1:

11110
11010
11000
00000
Answer: 1

Example 2:

11000
11000
00100
00011
Answer: 3

*/

class Solution {
    public int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    
    public int numIslands(char[][] grid) {
        if(grid.length == 0 || grid[0].length == 0) {
            return 0;
        }
        
        int m = grid.length;
        int n = grid[0].length;
        int count = 0;
        
        for(int r = 0 ; r < m ; r++) {
            for(int c = 0 ; c < n ; c++) {
                if(grid[r][c] == '1') {
                    count++;
                    dfs(grid, r, c, m, n);
                }
            }
        }
        return count;
    }
    
    public void dfs(char[][]grid, int r, int c, int m, int n) {
        if(r < 0 || r >= m || c < 0 || c >= n) { //out of bounds
            return;
        }
        
        for(int[] dir : dirs) {
            int dr = r + dir[0];
            int dc = c + dir[1];
            
            if(dr >= 0 && dr < m && dc >= 0 && dc < n && grid[dr][dc] == '1') { //if part of the original island it has already been counted and we can 0 it out
                grid[dr][dc] = '0';
                dfs(grid, dr, dc, m, n);
            }
        }
    }
}

File: Leetcode/200. Number of Islands/Solution.js
/* https://leetcode.com/problems/number-of-islands/

Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

Example 1:
Input: grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","0","0"]
]
Output: 1

Example 2:
Input: grid = [
  ["1","1","0","0","0"],
  ["1","1","0","0","0"],
  ["0","0","1","0","0"],
  ["0","0","0","1","1"]
]
Output: 3

Constraints:
m == grid.length
n == grid[i].length
1 <= m, n <= 300
grid[i][j] is '0' or '1'.
*/

/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    function dfs(r, c) {
        let dirs = [[0, -1], [0, 1], [-1, 0], [1, 0]];
        
        grid[r][c] = 0;
        dirs.forEach((dir) => {
            let dr = r+dir[0];
            let dc = c+dir[1];
            if(dr >= 0 && dr < m && dc >= 0 && dc < n && grid[dr][dc] === '1') {
                dfs(dr, dc);
            }
        })
    }
    
    let count = 0;
    let m = grid.length;
    let n = grid[0].length;
    
    for(let r = 0 ; r < m ; r++) {
        for(let c = 0 ; c < n ; c++) {
            if(grid[r][c] === '1') {
                count++;
                dfs(r, c);
            }
        }
    }
    return count;
};
